### ISSUE TITLE
BigDecimal - math operations [BUG]

### ISSUE THREAD

**Issue by nvasilov** [ta4j/ta4j#774]
1.102 + 0.0034 = 1.105; /* but the right value is  1.1054 */

```java
BigDecimal bigOne = BigDecimal.valueOf(1.102); /* precision 4 */
DecimalNum numOne = DecimalNum.valueOf(bigOne);

BigDecimal bigTwo = BigDecimal.valueOf(0.0034); /* precision 2 */
DecimalNum numTwo = DecimalNum.valueOf(bigTwo);

Num a1 /* 1.105 - it can be a problem (at least was for me) */ = numOne.plus(numTwo);

int maxPrecision = Math.max(bigOne.precision(), bigTwo.precision());
Num a2 /* 1.105 - get max precision is not enough */ = DecimalNum.valueOf(bigOne, maxPrecision).plus(numTwo);

Num a3 /* 1.1054 - to fix v1 */ = DecimalNum.valueOf(bigOne, maxPrecision + 1).plus(numTwo);
Num a4 /* 1.1054 - to fix v2 */ = DecimalNum.valueOf(bigOne.doubleValue()).plus(numTwo);

BigDecimal b /* 1.1054 */ = bigOne.add(bigTwo);
```

**Comment by MarcDahlem** [comment]
Try BigDecimal.valueOf("1.102") instead of BigDecimal.valueOf(1.102).  See https://stackoverflow.com/questions/43931672/bigdecimal-adding-wrong-value

**Comment by team172011** [comment]
@nvasilov good point. Maybe we should always use the default precision of `DecimalNum` (32) and do not let the user modify or set the precision for different instances of `DecimalNum` by using `valueOf` or the `DecimalNum(String val, int precision)` constructor ?

**Comment by nvasilov** [comment]
> @nvasilov good point. Maybe we should always use the default precision of `DecimalNum` (32) and do not let the user modify or set the precision for different instances of `DecimalNum` by using `valueOf` or the `DecimalNum(String val, int precision)` constructor ?

**Comment by nvasilov** [comment]
> @nvasilov good point. Maybe we should always use the default precision of `DecimalNum` (32) and do not let the user modify or set the precision for different instances of `DecimalNum` by using `valueOf` or the `DecimalNum(String val, int precision)` constructor ?

I agree, but if the user wants to set individual precision, that's their problem.

I think will be enough to delete the following method:
```java
    public static DecimalNum valueOf(BigDecimal val) {
        return new DecimalNum(val, val.precision());
    }
```

**Comment by team172011** [comment]
Deleting it or just change to:
```java
    public static DecimalNum valueOf(BigDecimal val) {
        return new DecimalNum(val, DEFAULT_PRECISION);
    }
```
?

**Comment by nimo23** [comment]
You can change the default to:

```
public static DecimalNum valueOf(BigDecimal val) {
    return new DecimalNum(val, DEFAULT_PRECISION);
}
```

and add another method with:

```
public static DecimalNum valueOf(BigDecimal val, int precision) {
    return new DecimalNum(val, precision);
}
```

Currently, the precision depends on the value coming from the source. For example, having a `barSeries` with prices of 4 decimal places, it will add a precision with 4 decimal places instead of 32. **Storing a value with 4 instead of 32 digits saves memory and is faster for computation.** Does it make sense to have always 32 digits when all the bars from a BarSeries only have at most a precision of 4? I think the DEFAULT_PRECISION should be the maximum allowed precision one could have, or? If user decides to set another (lower) precision, then it should be also allowed? 

The precision (decimal places) of the converted Num values is inherited by the Num-Function. For example, `series.function()` returns a Num-Function with a precision of 3, if their bar values have at most a precision of 3. I noticed that behaviour when creating https://github.com/ta4j/ta4j/pull/753 (this PR would also inherit its decimal places from its numType-example, which in this case leads to unintended values (`DecimalNum.ONE.function.valueOf(2.333)` uses the precision from the value instead of the default precision, which is not always the intended behaviour.)

**Comment by nimo23** [comment]
FYI: https://blogs.oracle.com/javamagazine/four-common-pitfalls-of-the-bigdecimal-class-and-how-to-avoid-them#anchor_2

**Comment by johnthetireless** [comment]
This is an old issue, but I will weigh in anyway,,  nimo is right.  32 significant digits is overkill for any TA use case I can think of.  Makes the objects bigger, but the real issue is performance.  I haven't tested this myself, but sources say the time for multiply and divide grows exponentially as the number of digits increases.  I know it gets slower to create BigDecimal objects.  I can't imagine sqrt() getting any faster.  My advice...  MathContext,Decimal64 is good enough for this domain.  Let people choose Decimal128 if they want; it will be less performant.  I have no reason to believe this, but the predefined contexts may be optimized.

**Comment by team172011** [comment]
merged #1016
